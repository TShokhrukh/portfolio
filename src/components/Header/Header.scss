.Header {
  --header-color: $dark;
  --header-container-background: #{$black};
  --header-container-gap: #{$spacer * 3};
  --header-container-padding-x: #{$spacer * 3};
  --header-container-padding-y: #{$spacer * 1.5};
  --header-container-border-radius: #{$spacer * 2};

  color: var(--header-color);
}

// .Header-Container {
//   border-radius: $spacer * 2;
//   background-color: $black;
//   display: flex;
//   gap: $spacer * 3;
//   color: $dark;
//   padding: $spacer * 1.5 $spacer * 3;
// }

// .Header-Nav {
//   display: flex;
//   align-items: center;
//   margin-left: auto;
// }
