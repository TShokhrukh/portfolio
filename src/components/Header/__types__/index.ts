export * from './Header.types'
export * from './Header-Nav.types'
