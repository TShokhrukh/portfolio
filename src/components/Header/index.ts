export * from './build'
export * from './__types__'
export * from './Header.viewModel'
export { Header as HeaderBase } from './Header'
