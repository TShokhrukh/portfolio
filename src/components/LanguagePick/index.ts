export * from './build'
export * from './__types__'
export * from './LanguagePick.viewModel'
export * from './LanguagePick.model'
export { LanguagePick as LanguagePickBase } from './LanguagePick'
