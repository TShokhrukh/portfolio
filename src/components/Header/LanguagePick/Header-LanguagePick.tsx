import React from 'react'
import { LanguagePick } from '~/components/LanguagePick/build'
import { HeaderContext } from '../libs/context'
import { cnHeader } from '../Header.const'

import type { IHeaderContext } from '../__types__'

export const HeaderLanguagePick = () => {
  const { viewModel } = React.useContext<IHeaderContext>(HeaderContext)

  return <LanguagePick viewModel={viewModel.language} className={cnHeader('LanguagePick')} />
}
