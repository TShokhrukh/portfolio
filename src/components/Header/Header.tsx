import React from 'react'
import { Drawer } from '~/components/Drawer'
import { Icon } from '~/components/Icon'
import { Button } from '~/components/Button'
import { cnHeader } from './Header.const'
import { HeaderContainer } from './Container'
import { HeaderNav } from './Nav'
import { HeaderLanguagePick } from './LanguagePick'
import { HeaderLogo } from './Logo'
import { HeaderContext } from './lib/context'
import './Header.scss'

import type { IHeaderProps } from './__types__'

export const Header: React.FC<IHeaderProps> = ({
  className,
  links,
  activeLink,
  viewModel,
  ...props
}) => {
  return (
    <HeaderContext.Provider value={{ links, activeLink, viewModel }}>
      <header {...props} className={cnHeader({}, [className])}>
        <HeaderContainer>
          <HeaderLogo />
          <div className={cnHeader('Navbar', { desktop: true })}>
            <HeaderNav direction='horizontal' />
          </div>
          <div className={cnHeader('Navbar', { mobile: true })}>
            <Drawer viewModel={viewModel.drawer} target={({ onClick }) => (
              <Button onClick={onClick} className={cnHeader('BarsBtn', {})}>
                <Icon name='bars' />
              </Button>
            )} title={<HeaderLogo />}>
              <HeaderNav direction='vertical' onClick={viewModel.drawer.close} />
            </Drawer>
          </div>
          <HeaderLanguagePick />
        </HeaderContainer>
      </header>
    </HeaderContext.Provider>
  )
}
