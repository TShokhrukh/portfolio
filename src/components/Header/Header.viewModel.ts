import { LanguagePickViewModel, ILanguagePickModel, ILanguagePickViewModel } from '~/components/LanguagePick'
import { DrawerViewMode, IDrawerModel, IDrawerViewModel } from '~/components/Drawer'
import type { IPopperModel } from '~/components/Popper'
import type { IHeaderViewModel } from './__types__'

export class HeaderViewModel implements IHeaderViewModel {
  language: ILanguagePickViewModel
  drawer: IDrawerViewModel

  constructor (
    languageModel: ILanguagePickModel,
    popperModel: IPopperModel,
    drawerModel: IDrawerModel
  ) {
    this.language = new LanguagePickViewModel(languageModel, popperModel)
    this.drawer = new DrawerViewMode(drawerModel)
  }
}
