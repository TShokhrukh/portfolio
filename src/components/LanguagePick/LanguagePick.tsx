import React, { useRef } from 'react'
import { Observer } from 'mobx-react'
import { Text } from '@eo-locale/react'
import { Button } from '../Button/Button'
import { Popper } from '../Popper/Popper'
import { cnLanguagePick } from './LanguagePick.const'

import type { ILanguagePickProps } from './__types__'

export const LanguagePick: React.FC<ILanguagePickProps> = ({
  id,
  viewModel,
  className,
  ...props
}) => {
  const anchorEl = useRef<HTMLElement>(null)
  return (
    <Popper id={id} viewModel={viewModel} anchorEl={anchorEl}
      target={(targetProps) => (
        <Button {...props} {...targetProps} innerRef={anchorEl} className={cnLanguagePick({}, [className])}>
          <Observer>
            {() => <Text id={`language.${viewModel.current}`} />}
          </Observer>
        </Button>
      )}>
      { viewModel.supported.map(lang => (
        <Button key={lang} onClick={() => viewModel.setLocal(lang)}>
          <Text id={`language.${lang}`} />
        </Button>
      )) }
    </Popper>
  )
}

LanguagePick.defaultProps = {
  id: cnLanguagePick()
}
