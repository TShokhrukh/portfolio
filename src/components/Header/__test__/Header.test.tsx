/* global test, expect, describe, beforeEach */
import React from 'react'
import { render, screen } from '@testing-library/react'
import { HeaderWrapper } from './HeaderWrapper'
import { Header } from '../Header'
import { HeaderViewModel } from '../Header.viewModel'
import { LanguagePickModel } from '../../LanguagePick/LanguagePick.model'
import { PopperModel } from '../../Popper/Popper.model'

import type { IHeaderViewModel } from '../__types__'
import type { ILink } from '../../Nav/__types__'

describe('Header', () => {
  let viewModel: IHeaderViewModel

  beforeEach(() => {
    viewModel = new HeaderViewModel(
      new LanguagePickModel('en', ['en', 'ru', 'uz']),
      new PopperModel()
    )
  })

  test('should return the component template', () => {
    render(
      <HeaderWrapper>
        <Header viewModel={viewModel} links={[]} data-testid='header' />
      </HeaderWrapper>
    )

    expect(screen.getByTestId('header')).toMatchSnapshot()
  })

  test('should render links', () => {
    const links: ILink[] = [
      {
        href: 'one',
        label: 'link-1'
      },
      {
        href: 'two',
        label: 'link-2'
      }
    ]
    render(
      <HeaderWrapper>
        <Header viewModel={viewModel} links={links} data-testid='header' />
      </HeaderWrapper>
    )

    expect(screen.queryByText('link-1')).toBeInTheDocument()
    expect(screen.queryByText('link-1')).toHaveAttribute('href', 'one')
    expect(screen.queryByText('link-2')).toBeInTheDocument()
    expect(screen.queryByText('link-2')).toHaveAttribute('href', 'two')
  })
})
