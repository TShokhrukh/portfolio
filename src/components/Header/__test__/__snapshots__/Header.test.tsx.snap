// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header must return the component template 1`] = `
<header
  class="Header"
  data-testid="header"
>
  <div
    class="Container Header-Container"
  >
    <div
      class="Header-Logo"
    >
      <img
        alt="..."
        class="Header-LogoImage"
      />
    </div>
    <div
      class="Header-Navbar Header-Navbar_desktop"
    >
      <nav
        class="Nav Nav_direction_horizontal Header-Nav"
      >
        <ul
          class="Nav-List"
        />
      </nav>
    </div>
    <div
      class="Header-Navbar Header-Navbar_mobile"
    >
      <button
        class="Button Header-BarsBtn"
        type="button"
      >
        <span
          aria-hidden="true"
          class="fa fa-bars Icon"
        />
      </button>
    </div>
    <button
      aria-describedby="LanguagePick"
      class="Button LanguagePick Header-LanguagePick"
      type="button"
    >
      en
    </button>
  </div>
</header>
`;
